tools/autograd/gen_variable_type.py

```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_orignal_self_definition() -> List[str]:
+        body: List[str] = []
+        if inplace:
+            body.append('c10::optional<at::Tensor> original_self;')
```

Comment:
I remember Ed saying something a while ago about splitting off forward-mode AD into its own DispatchKey (to be separate from reverse-mode AD). Some of the optimizations we're doing here (clone only once if both forward mode and reverse mode need a close of the input) are difficult to do if we do split forward-mode AD into different dispatch keys. Is this a concern?

Reply:
We might have to give up such optimizations when we move to a different dispatch key indeed.
We might be able to unpack the SavedVariable in the Node we just created, but that would be more complex than the current code for sure.